File: src/net/sf/image4j/codec/bmp/BMPDecoder.java
/*
 * Decodes a BMP image from an <tt>InputStream</tt> to a <tt>BufferedImage</tt>
 *
 * @author Ian McDonagh
 */

package net.sf.image4j.codec.bmp;

import me.anno.image.Image;
import me.anno.image.raw.IntImage;
import net.sf.image4j.io.CountingInputStream;
import net.sf.image4j.io.LittleEndianInputStream;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;

/**
 * Decodes images in BMP format.
 *
 * @author Ian McDonagh
 */
public class BMPDecoder {

    private final Image img;

    /**
     * Creates a new instance of BMPDecoder and reads the BMP data from the source.
     *
     * @param in the source <tt>InputStream</tt> from which to read the BMP data
     * @throws java.io.IOException if an error occurs
     */
    public BMPDecoder(InputStream in) throws IOException {
        LittleEndianInputStream lis = new LittleEndianInputStream(new CountingInputStream(in));

        /* header [14] */

        //signature "BM" [2]
        byte[] signatureBytes = new byte[2];
        lis.read(signatureBytes);
        String signature = new String(signatureBytes, StandardCharsets.UTF_8);

        if (!signature.equals("BM")) {
            throw new IOException("Invalid signature '" + signature + "' for BMP format");
        }

        // file size [4]
        lis.readIntLE();

        // reserved = 0 [4]
        lis.readIntLE();

        //DataOffset [4] file offset to raster data
        lis.readIntLE();

        /* info header [40] */

        InfoHeader infoHeader = readInfoHeader(lis);

        /* Color table and Raster data */

        img = read(infoHeader, lis);
    }

    /**
     * Retrieves a bit from the lowest order byte of the given integer.
     *
     * @param bits  the source integer, treated as an unsigned byte
     * @param index the index of the bit to retrieve, which must be in the range <tt>0..7</tt>.
     * @return the bit at the specified index, which will be either <tt>0</tt> or <tt>1</tt>.
     */
    private static int getBit(int bits, int index) {
        return (bits >> (7 - index)) & 1;
    }

    /**
     * Retrieves a nibble (4 bits) from the lowest order byte of the given integer.
     *
     * @param nibbles the source integer, treated as an unsigned byte
     * @param index   the index of the nibble to retrieve, which must be in the range <tt>0..1</tt>.
     * @return the nibble at the specified index, as an unsigned byte.
     */
    private static int getNibble(int nibbles, int index) {
        return (nibbles >> (4 * (1 - index))) & 0xF;
    }

    /**
     * The decoded image read from the source input.
     *
     * @return the <tt>BufferedImage</tt> representing the BMP image.
     */
    public Image getImage() {
        return img;
    }

    /**
     * Reads the BMP info header structure from the given <tt>InputStream</tt>.
     *
     * @param lis the <tt>InputStream</tt> to read
     * @return the <tt>InfoHeader</tt> structure
     * @throws java.io.IOException if an error occurred
     */
    public static InfoHeader readInfoHeader(net.sf.image4j.io.LittleEndianInputStream lis) throws IOException {
        return new InfoHeader(lis);
    }

    /**
     * @since 0.6
     */
    public static InfoHeader readInfoHeader(net.sf.image4j.io.LittleEndianInputStream lis, int infoSize) throws IOException {
        return new InfoHeader(lis, infoSize);
    }

    /**
     * Reads the BMP data from the given <tt>InputStream</tt> using the information
     * contained in the <tt>InfoHeader</tt>.
     *
     * @param lis        the source input
     * @param infoHeader an <tt>InfoHeader</tt> that was read by a call to
     *                   {@link #readInfoHeader(net.sf.image4j.io.LittleEndianInputStream) readInfoHeader()}.
     * @return the decoded image read from the source input
     * @throws java.io.IOException if an error occurs
     */
    public static IntImage read(InfoHeader infoHeader, net.sf.image4j.io.LittleEndianInputStream lis) throws IOException {

        /* Color table (palette) */
        int[] colorTable = null;

        // color table is only present for 1, 4 or 8 bit (indexed) images
        if (infoHeader.sBitCount <= 8) {
            colorTable = readColorTable(infoHeader, lis);
        }

        return read(infoHeader, lis, colorTable);

    }

    /**
     * Reads the BMP data from the given <tt>InputStream</tt> using the information
     * contained in the <tt>InfoHeader</tt>.
     *
     * @param colorTable <tt>ColorEntry</tt> array containing palette
     * @param infoHeader an <tt>InfoHeader</tt> that was read by a call to
     *                   {@link #readInfoHeader(net.sf.image4j.io.LittleEndianInputStream) readInfoHeader()}.
     * @param lis        the source input
     * @return the decoded image read from the source input
     * @throws java.io.IOException if any error occurs
     */
    public static IntImage read(InfoHeader infoHeader, LittleEndianInputStream lis, int[] colorTable) throws IOException {
        if (infoHeader.sBitCount == 1 && infoHeader.iCompression == BMPConstants.BI_RGB) {// 1-bit (monochrome) uncompressed
            return read1(infoHeader, lis, colorTable);
        } else if (infoHeader.sBitCount == 4 && infoHeader.iCompression == BMPConstants.BI_RGB) {// 4-bit uncompressed
            return read4(infoHeader, lis, colorTable);
        } else if (infoHeader.sBitCount == 8 && infoHeader.iCompression == BMPConstants.BI_RGB) { // 8-bit uncompressed
            return read8(infoHeader, lis, colorTable);
        } else if (infoHeader.sBitCount == 24 && infoHeader.iCompression == BMPConstants.BI_RGB) {// 24-bit uncompressed
            return read24(infoHeader, lis);
        } else if (infoHeader.sBitCount == 32 && infoHeader.iCompression == BMPConstants.BI_RGB) {// 32bit uncompressed
            return read32(infoHeader, lis);
        } else {
            throw new IOException("Unrecognized bitmap format: bit count=" + infoHeader.sBitCount + ", compression=" +
                    infoHeader.iCompression);
        }
    }

    /**
     * Reads the <tt>ColorEntry</tt> table from the given <tt>InputStream</tt> using
     * the information contained in the given <tt>infoHeader</tt>.
     *
     * @param infoHeader the <tt>InfoHeader</tt> structure, which was read using
     *                   {@link #readInfoHeader(net.sf.image4j.io.LittleEndianInputStream) readInfoHeader()}
     * @param lis        the <tt>InputStream</tt> to read
     * @return the decoded image read from the source input
     * @throws java.io.IOException if an error occurs
     */
    public static int[] readColorTable(InfoHeader infoHeader, LittleEndianInputStream lis) throws IOException {
        int[] colorTable = new int[infoHeader.iNumColors];
        for (int i = 0; i < infoHeader.iNumColors; i++) {
            colorTable[i] = lis.readIntLE();
        }
        return colorTable;
    }

    /**
     * Reads 1-bit uncompressed bitmap raster data, which may be monochrome depending on the
     * palette entries in <tt>colorTable</tt>.
     *
     * @param infoHeader the <tt>InfoHeader</tt> structure, which was read using
     *                   {@link #readInfoHeader(net.sf.image4j.io.LittleEndianInputStream) readInfoHeader()}
     * @param lis        the source input
     * @param colors     int[] specifying the palette, which
     *                   must not be <tt>null</tt>.
     * @return the decoded image read from the source input
     * @throws java.io.IOException if an error occurs
     */
    public static IntImage read1(InfoHeader infoHeader, LittleEndianInputStream lis, int[] colors) throws IOException {
        //1 bit per pixel or 8 pixels per byte
        //each pixel specifies the palette index

        // padding
        int bitsPerLine = infoHeader.iWidth;
        if (bitsPerLine % 32 != 0) {
            bitsPerLine = (bitsPerLine / 32 + 1) * 32;
        }

        int bytesPerLine = bitsPerLine / 8;

        int[] line = new int[bytesPerLine];
        int[] data = new int[infoHeader.iWidth * infoHeader.iHeight];
        for (int y = infoHeader.iHeight - 1; y >= 0; y--) {
            for (int i = 0; i < bytesPerLine; i++) {
                line[i] = lis.readUnsignedByte();
            }

            int ctr = y * infoHeader.iWidth;
            for (int x = 0; x < infoHeader.iWidth; x++) {
                int i = x >> 3;
                int v = line[i];
                int b = x & 7;
                int index = getBit(v, b);
                // img.setRGB(x, y, rgb);
                // set the sample (colour index) for the pixel
                data[ctr++] = colors[index];
            }
        }

        return new IntImage(infoHeader.iWidth, infoHeader.iHeight, data, false);
    }

    /**
     * Reads 4-bit uncompressed bitmap raster data, which is interpreted based on the colours
     * specified in the palette.
     *
     * @param infoHeader the <tt>InfoHeader</tt> structure, which was read using
     *                   {@link #readInfoHeader(net.sf.image4j.io.LittleEndianInputStream) readInfoHeader()}
     * @param lis        the source input
     * @param colors     int[] specifying the palette, which
     *                   must not be <tt>null</tt>.
     * @return the decoded image read from the source input
     * @throws java.io.IOException if an error occurs
     */
    public static IntImage read4(InfoHeader infoHeader, LittleEndianInputStream lis, int[] colors) throws IOException {

        // 2 pixels per byte or 4 bits per pixel.
        // Colour for each pixel specified by the color index in the palette.

        //padding
        int bitsPerLine = infoHeader.iWidth * 4;
        if (bitsPerLine % 32 != 0) {
            bitsPerLine = (bitsPerLine / 32 + 1) * 32;
        }
        int bytesPerLine = bitsPerLine / 8;

        int[] line = new int[bytesPerLine];
        int[] data = new int[infoHeader.iWidth * infoHeader.iHeight];
        for (int y = infoHeader.iHeight - 1; y >= 0; y--) {

            // scan line
            for (int i = 0; i < bytesPerLine; i++) {
                int b = lis.readUnsignedByte();
                line[i] = b;
            }

            // get pixels
            int ctr = y * infoHeader.iWidth;
            for (int x = 0; x < infoHeader.iWidth; x++) {
                //get byte index for line
                int b = x >> 1; // 2 pixels per byte
                int i = x & 1;
                int n = line[b];
                int index = getNibble(n, i);
                data[ctr++] = colors[index];
            }
        }

        return new IntImage(infoHeader.iWidth, infoHeader.iHeight, data, false);
    }

    /**
     * Reads 8-bit uncompressed bitmap raster data, which is interpreted based on the colours
     * specified in the palette.
     *
     * @param infoHeader the <tt>InfoHeader</tt> structure, which was read using
     *                   {@link #readInfoHeader(net.sf.image4j.io.LittleEndianInputStream) readInfoHeader()}
     * @param lis        the source input
     * @param colors int[] specifying the palette, which
     *                   must not be <tt>null</tt>.
     * @return the decoded image read from the source input
     * @throws java.io.IOException if an error occurs
     */
    public static IntImage read8(InfoHeader infoHeader, LittleEndianInputStream lis, int[] colors) throws IOException {

        // 1 byte per pixel
        //  color index 1 (index of color in palette)
        // lines padded to nearest 32bits
        // no alpha

        //padding
        int dataPerLine = infoHeader.iWidth;
        int bytesPerLine = dataPerLine;
        if (bytesPerLine % 4 != 0) {
            bytesPerLine = (bytesPerLine / 4 + 1) * 4;
        }
        int padBytesPerLine = bytesPerLine - dataPerLine;

        int[] data = new int[infoHeader.iWidth * infoHeader.iHeight];
        for (int y = infoHeader.iHeight - 1; y >= 0; y--) {
            int ctr = y * infoHeader.iWidth;
            for (int x = 0; x < infoHeader.iWidth; x++) {
                int b = lis.readUnsignedByte();
                data[ctr++] = colors[b];
            }
            lis.skip(padBytesPerLine);
        }

        return new IntImage(infoHeader.iWidth, infoHeader.iHeight, data, false);
    }

    /**
     * Reads 24-bit uncompressed bitmap raster data.
     *
     * @param lis        the source input
     * @param infoHeader the <tt>InfoHeader</tt> structure, which was read using
     *                   {@link #readInfoHeader(net.sf.image4j.io.LittleEndianInputStream) readInfoHeader()}
     * @return the decoded image read from the source input
     * @throws java.io.IOException if an error occurs
     */
    public static IntImage read24(InfoHeader infoHeader, LittleEndianInputStream lis) throws IOException {

        // 3 bytes per pixel
        //  blue 1
        //  green 1
        //  red 1
        // lines padded to nearest 32 bits
        // no alpha

        // padding to nearest 32 bits
        int dataPerLine = infoHeader.iWidth * 3;
        int bytesPerLine = dataPerLine;
        if (bytesPerLine % 4 != 0) {
            bytesPerLine = (bytesPerLine / 4 + 1) * 4;
        }
        int padBytesPerLine = bytesPerLine - dataPerLine;

        int[] data = new int[infoHeader.iHeight * infoHeader.iWidth];
        for (int y = infoHeader.iHeight - 1; y >= 0; y--) {
            int ctr = y * infoHeader.iWidth;
            for (int x = 0; x < infoHeader.iWidth; x++) {
                int b = lis.readUnsignedByte();
                int g = lis.readUnsignedByte();
                int r = lis.readUnsignedByte();
                data[ctr++] = (r << 16) | (g << 8) | b;
            }
            lis.skip(padBytesPerLine);
        }

        return new IntImage(infoHeader.iWidth, infoHeader.iHeight, data, false);

    }


    /**
     * Reads 32-bit uncompressed bitmap raster data, with transparency.
     *
     * @param lis        the source input
     * @param infoHeader the <tt>InfoHeader</tt> structure, which was read using
     *                   {@link #readInfoHeader(net.sf.image4j.io.LittleEndianInputStream) readInfoHeader()}
     * @return the decoded image read from the source input
     * @throws java.io.IOException if an error occurs
     */
    public static IntImage read32(InfoHeader infoHeader, LittleEndianInputStream lis) throws IOException {
        // 4 bytes per pixel
        // blue 1
        // green 1
        // red 1
        // alpha 1
        // No padding since each pixel = 32 bits

        int[] data = new int[infoHeader.iWidth * infoHeader.iHeight];
        for (int y = infoHeader.iHeight - 1; y >= 0; y--) {
            int ctr = y * infoHeader.iWidth;
            for (int x = 0; x < infoHeader.iWidth; x++) {
                data[ctr++] = lis.readIntLE();// bgra
            }
        }

        return new IntImage(infoHeader.iWidth, infoHeader.iHeight, data, true);
    }

    /**
     * Reads and decodes BMP data from the source input.
     *
     * @param in the source input
     * @return the decoded image read from the source file
     * @throws java.io.IOException if an error occurs
     */
    public static Image read(InputStream in) throws IOException {
        BMPDecoder d = new BMPDecoder(in);
        return d.getImage();
    }

}

File: src/net/sf/image4j/codec/bmp/BMPDecoder.kt
package net.sf.image4j.codec.bmp

import me.anno.image.Image
import me.anno.image.raw.IntImage
import me.anno.io.Streams.readLE32
import net.sf.image4j.io.CountingInputStream
import net.sf.image4j.io.Utils
import java.io.EOFException
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets

/**
 * Decodes images in BMP format.
 * Creates a new instance of BMPDecoder and reads the BMP data from the source.
 *
 * @param input the source <tt>InputStream</tt> from which to read the BMP data
 * @throws IOException if an error occurs
 * @author Ian McDonagh
 */
class BMPDecoder(input: InputStream) {

    companion object {

        /**
         * no compression
         * */
        const val BI_RGB = 0
        const val BI_RLE8 = 1
        const val BI_RLE4 = 2

        /**
         * Retrieves a bit from the lowest order byte of the given integer.
         *
         * @param bits  the source integer, treated as an unsigned byte
         * @param index the index of the bit to retrieve, which must be in the range <tt>0..7</tt>.
         * @return the bit at the specified index, which will be either <tt>0</tt> or <tt>1</tt>.
         */
        @JvmStatic
        private fun getBit(bits: Int, index: Int): Int {
            return (bits shr (7 - index)) and 1
        }

        /**
         * Retrieves a nibble (4 bits) from the lowest order byte of the given integer.
         *
         * @param nibbles the source integer, treated as an unsigned byte
         * @param index   the index of the nibble to retrieve, which must be in the range <tt>0..1</tt>.
         * @return the nibble at the specified index, as an unsigned byte.
         */
        @JvmStatic
        private fun getNibble(nibbles: Int, index: Int): Int {
            return nibbles shr 4 * (1 - index) and 0xf
        }

        @JvmStatic
        fun readInfoHeader(lis: CountingInputStream) =
            InfoHeader(lis)

        @JvmStatic
        fun readInfoHeader(lis: CountingInputStream, infoSize: Int) =
            InfoHeader(lis, infoSize)

        /**
         * Reads the BMP data from the given <tt>InputStream</tt> using the information
         * contained in the <tt>InfoHeader</tt>.
         *
         * @param lis        the source input
         * @param infoHeader an <tt>InfoHeader</tt> that was read by a call to
         * [readInfoHeader()][.readInfoHeader].
         * @return the decoded image read from the source input
         * @throws java.io.IOException if an error occurs
         */
        @JvmStatic
        fun read(infoHeader: InfoHeader, lis: CountingInputStream): IntImage {
            // Color table (palette)
            var colorTable: IntArray? = null
            // color table is only present for 1, 4 or 8 bit (indexed) images
            if (infoHeader.bitCount <= 8) {
                colorTable = readColorTable(infoHeader, lis)
            }
            return read(infoHeader, lis, colorTable)
        }

        /**
         * Reads the BMP data from the given <tt>InputStream</tt> using the information
         * contained in the <tt>InfoHeader</tt>.
         *
         * @param colorTable <tt>ColorEntry</tt> array containing palette
         * @param infoHeader an <tt>InfoHeader</tt> that was read by a call to
         * [readInfoHeader()][.readInfoHeader].
         * @param lis        the source input
         * @return the decoded image read from the source input
         * @throws java.io.IOException if any error occurs
         */
        @JvmStatic
        fun read(infoHeader: InfoHeader, lis: CountingInputStream, colorTable: IntArray?): IntImage {
            val image = if (infoHeader.compression == BI_RGB) {
                when (infoHeader.bitCount) {
                    1 -> read1(infoHeader, lis, colorTable!!) // 1-bit (monochrome) uncompressed
                    4 -> read4(infoHeader, lis, colorTable!!) // 4-bit uncompressed
                    8 -> read8(infoHeader, lis, colorTable!!) // 8-bit uncompressed
                    24 -> read24(infoHeader, lis) // 24-bit uncompressed
                    32 -> read32(infoHeader, lis) // 32bit uncompressed
                    else -> null
                }
            } else null
            return image
                ?: throw IOException("Unrecognized bitmap format: bit count=${infoHeader.bitCount}, compression=${infoHeader.compression}")
        }

        @JvmStatic
        fun readColorTable(infoHeader: InfoHeader, lis: CountingInputStream) =
            IntArray(infoHeader.numColors) { lis.readLE32() }

        /**
         * Reads 1-bit uncompressed bitmap raster data, which may be monochrome depending on the
         * palette entries in <tt>colorTable</tt>.
         */
        @JvmStatic
        fun read1(infoHeader: InfoHeader, lis: CountingInputStream, colors: IntArray): IntImage {
            //1 bit per pixel or 8 pixels per byte
            //each pixel specifies the palette index

            // padding
            var bitsPerLine = infoHeader.width
            if (bitsPerLine % 32 != 0) {
                bitsPerLine = (bitsPerLine / 32 + 1) * 32
            }
            val bytesPerLine = bitsPerLine / 8
            val line = IntArray(bytesPerLine)
            val data = IntArray(infoHeader.width * infoHeader.height)
            for (y in infoHeader.height - 1 downTo 0) {
                for (i in 0 until bytesPerLine) {
                    line[i] = lis.read()
                }
                var ctr = y * infoHeader.width
                for (x in 0 until infoHeader.width) {
                    val i = x shr 3
                    val v = line[i]
                    val b = x and 7
                    val index = getBit(v, b)
                    // img.setRGB(x, y, rgb);
                    // set the sample (colour index) for the pixel
                    data[ctr++] = colors[index]
                }
            }
            return IntImage(infoHeader.width, infoHeader.height, data, false)
        }

        /**
         * Reads 4-bit uncompressed bitmap raster data, which is interpreted based on the colours
         * specified in the palette.
         *
         * @param infoHeader the <tt>InfoHeader</tt> structure, which was read using
         * [readInfoHeader()][.readInfoHeader]
         * @param lis        the source input
         * @param colors     int[] specifying the palette, which
         * must not be <tt>null</tt>.
         * @return the decoded image read from the source input
         * @throws IOException if an error occurs
         */
        @JvmStatic
        fun read4(infoHeader: InfoHeader, lis: CountingInputStream, colors: IntArray): IntImage {

            // 2 pixels per byte or 4 bits per pixel.
            // Colour for each pixel specified by the color index in the palette.

            //padding
            var bitsPerLine = infoHeader.width * 4
            if (bitsPerLine % 32 != 0) {
                bitsPerLine = (bitsPerLine / 32 + 1) * 32
            }
            val bytesPerLine = bitsPerLine / 8
            val line = IntArray(bytesPerLine)
            val data = IntArray(infoHeader.width * infoHeader.height)
            for (y in infoHeader.height - 1 downTo 0) {

                // scan line
                for (i in 0 until bytesPerLine) {
                    line[i] = lis.read()
                }

                // get pixels
                var ctr = y * infoHeader.width
                for (x in 0 until infoHeader.width) {
                    //get byte index for line
                    val b = x shr 1 // 2 pixels per byte
                    val i = x and 1
                    val n = line[b]
                    val index = getNibble(n, i)
                    data[ctr++] = colors[index]
                }
            }
            return IntImage(infoHeader.width, infoHeader.height, data, false)
        }

        /**
         * Reads 8-bit uncompressed bitmap raster data, which is interpreted based on the colours
         * specified in the palette.
         *
         * @param infoHeader the <tt>InfoHeader</tt> structure, which was read using
         * [readInfoHeader()][.readInfoHeader]
         * @param lis        the source input
         * @param colors     int[] specifying the palette, which
         * must not be <tt>null</tt>.
         * @return the decoded image read from the source input
         * @throws IOException if an error occurs
         */
        @JvmStatic
        fun read8(infoHeader: InfoHeader, lis: CountingInputStream, colors: IntArray): IntImage {

            // 1 byte per pixel
            //  color index 1 (index of color in palette)
            // lines padded to nearest 32bits
            // no alpha

            //padding
            val dataPerLine = infoHeader.width
            var bytesPerLine = dataPerLine
            if (bytesPerLine % 4 != 0) {
                bytesPerLine = (bytesPerLine / 4 + 1) * 4
            }
            val padBytesPerLine = bytesPerLine - dataPerLine
            val width = infoHeader.width
            val data = IntArray(width * infoHeader.height)
            for (y in infoHeader.height - 1 downTo 0) {
                var ctr = y * width
                for (x in 0 until width) {
                    data[ctr++] = colors[lis.read()]
                }
                Utils.skip(lis, padBytesPerLine, false)
            }
            return IntImage(infoHeader.width, infoHeader.height, data, false)
        }

        /**
         * Reads 24-bit uncompressed bitmap raster data.
         *
         * @param lis        the source input
         * @param infoHeader the <tt>InfoHeader</tt> structure, which was read using
         * [readInfoHeader()][.readInfoHeader]
         * @return the decoded image read from the source input
         * @throws IOException if an error occurs
         */
        @JvmStatic
        fun read24(infoHeader: InfoHeader, lis: CountingInputStream): IntImage {

            // 3 bytes per pixel
            //  blue 1
            //  green 1
            //  red 1
            // lines padded to nearest 32 bits
            // no alpha

            // padding to nearest 32 bits
            val dataPerLine = infoHeader.width * 3
            var bytesPerLine = dataPerLine
            if (bytesPerLine % 4 != 0) {
                bytesPerLine = (bytesPerLine / 4 + 1) * 4
            }
            val padBytesPerLine = bytesPerLine - dataPerLine
            val width = infoHeader.width
            val data = IntArray(width * infoHeader.height)
            for (y in infoHeader.height - 1 downTo 0) {
                var ctr = y * width
                for (x in 0 until width) {
                    val b = lis.read()
                    val g = lis.read()
                    val r = lis.read()
                    data[ctr++] = r shl 16 or (g shl 8) or b
                }
                Utils.skip(lis, padBytesPerLine, false)
            }
            return IntImage(infoHeader.width, infoHeader.height, data, false)
        }

        /**
         * Reads 32-bit uncompressed bitmap raster data, with transparency.
         *
         * @param lis        the source input
         * @param infoHeader the <tt>InfoHeader</tt> structure, which was read using
         * [readInfoHeader()][.readInfoHeader]
         * @return the decoded image read from the source input
         * @throws IOException if an error occurs
         */
        @JvmStatic
        fun read32(infoHeader: InfoHeader, lis: CountingInputStream): IntImage {
            // 4 bytes per pixel
            // blue 1
            // green 1
            // red 1
            // alpha 1
            // No padding since each pixel = 32 bits
            val data = IntArray(infoHeader.width * infoHeader.height)
            for (y in infoHeader.height - 1 downTo 0) {
                var ctr = y * infoHeader.width
                for (x in 0 until infoHeader.width) {
                    data[ctr++] = lis.readLE32() // bgra
                }
            }
            return IntImage(infoHeader.width, infoHeader.height, data, true)
        }

        /**
         * Reads and decodes BMP data from the source input.
         *
         * @param `in` the source input
         * @return the decoded image read from the source file
         * @throws IOException if an error occurs
         */
        @JvmStatic
        fun read(input: InputStream) = BMPDecoder(input).image
    }

    /**
     * The decoded image read from the source input.
     *
     * @return the <tt>Image</tt> representing the BMP image.
     */
    val image: Image

    init {

        val lis = CountingInputStream(CountingInputStream(input))

        /* header [14] */

        // signature "BM" [2]
        val s0 = lis.read()
        val s1 = lis.read()
        if (s0 < 0 || s1 < 0) throw EOFException()
        if (s0 != 'B'.code || s1 != 'M'.code) {
            val signature = String(byteArrayOf(s0.toByte(), s1.toByte()), StandardCharsets.UTF_8)
            throw IOException("Invalid signature '$signature' for BMP format")
        }

        // file size [4]
        lis.readLE32()

        // reserved = 0 [4]
        lis.readLE32()

        //DataOffset [4] file offset to raster data
        lis.readLE32()

        /* info header [40] */
        val infoHeader = readInfoHeader(lis)

        /* Color table and Raster data */
        image = read(infoHeader, lis)

    }
}